### ISSUE TITLE
Middleware that return early break `Router::or`

### ISSUE THREAD

**Issue by Flowneee** [tokio-rs/axum#380]
### Version

```
├── axum v0.2.8
```

### Platform

Linux fdr12.tcsbank.ru 5.14.9-200.fc34.x86_64 #1 SMP Thu Sep 30 11:55:35 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Description

When I try to split app into multiple Routers (like in docs https://docs.rs/axum/0.2.8/axum/#to-groups-of-routes), layers from "leaf" Routers affect other "leaf" Routers.

I tried this code:

```rust
use axum::{
    handler::{get, post, Handler},
    response::{Html, IntoResponse},
    Router,
};
use hyper::StatusCode;
use std::net::SocketAddr;
use tower_http::auth::RequireAuthorizationLayer;

async fn handler_404() -> impl IntoResponse {
    StatusCode::NOT_FOUND
}

#[tokio::main]
async fn main() {
    let foo = Router::new()
        .route("/", get(|| async {}))
        .route("/foo", post(|| async {}));

    let bar = Router::new()
        .route("/requires-auth", get(|| async {}))
        .layer(RequireAuthorizationLayer::basic("a", "a"));

    let app = bar.or(foo).or(handler_404.into_service());  # NOTE: bar is first, not foo

    let addr = "0.0.0.0:8080".parse::<SocketAddr>().unwrap();
    axum::Server::try_bind(&addr)
        .unwrap()
        .serve(app.into_make_service())
        .await;
}
```

I expected to see this happen: when I do `curl localhost:8080/`, I get `200 OK`

Instead, this happened: `401 Unauthorized`.

Basically now all routes require auth from `bar` "leaf" Router.

If I apply auth layer to handler directly, everything works, but this mean that I cannot apply layer to group of routes, only to all of them at once.

I think that layers should apply only to current Router, at least this is how I understand current documentation.

**Comment by davidpdrsn** [comment]
> #363 won't fix this problem?

I don't believe so.

> Don't you think that current documentation is misleading then?

You're right, we probably shouldn't use the particular example we do since it wont work.

**Comment by davidpdrsn** [comment]
I'm not sure. Made a note to look into it. You're welcome to test it if you want.

**Comment by Flowneee** [comment]
I think renaming `or` to `merge` _kind of_ solves problem, because (at least for me) semantically `merge`is more like concatenation of two lists, and that is what `merge` do exactly - it make single list from 2 lists of `Service`es (which Router is roughly), and `or` is more like creating tree-like structure. `nest` is also creating tree IMO, which was slightly confusing.

But I think it would be cool to have something like `nest`, but with non-unique prefix, so you can mount 2 Routers under single prefix and processing request would be like this:
- check routes in first Router. If successful - proceed with first Router, if not - next step;
- check routes in second Router. If successful - proceed with second Router, if not - repeat until you run out of Routers with current prefix;
- if none matched, proceed as 404 or whatever default route is.

This will allow to create tree-like structure of Routers, where layers can be applied only to certain routes, which _do not_ share common prefix.

**Comment by davidpdrsn** [comment]
How is that different from `.nest("/foo", router_a.merge(router_b))`?

**Comment by Flowneee** [comment]
Layers from both routers will mix again. What I think is something like

```rust
let router_a = Router::new().route("/metrics", ...).route("/health", ...);
let router_b = Router::new().route("/shutdown", ...).route("/kill", ...).layer(SomeAuthLayer {});

let router = Route::new().nest("/manage", router_a).nest("/manage", router_b);
```
Ofc there is always option to apply layers to handlers directly, even though it is verbose.

**Comment by davidpdrsn** [comment]
> Layers from both routers will mix again

What do mean by this? Layers only apply to routes added to the router at the time the layer is added.

**Comment by Flowneee** [comment]
Seems I got it wrong. If replace `or` with `merge`, on new axum version code from first message works as expected

```rust
use axum::{
    handler::Handler,
    response::IntoResponse,
    routing::{get, post},
    Router,
};
use hyper::StatusCode;
use std::net::SocketAddr;
use tower_http::auth::RequireAuthorizationLayer;

async fn handler_404() -> impl IntoResponse {
    StatusCode::NOT_FOUND
}

#[tokio::main]
async fn main() {
    let foo = Router::new()
        .route("/", get(|| async {}))
        .route("/foo", post(|| async {}));

    let bar = Router::new()
        .route("/requires-auth", get(|| async {}))
        .layer(RequireAuthorizationLayer::basic("a", "a"));

    let app = bar.merge(foo).fallback(handler_404.into_service()); //NOTE: bar is first, not foo

    let addr = "0.0.0.0:8080".parse::<SocketAddr>().unwrap();
    axum::Server::try_bind(&addr)
        .unwrap()
        .serve(app.into_make_service())
        .await;
}
```

Thanks!
